### DISCUSSION TITLE
Fetching random images from the input folder

### DISCUSSION THREAD

**Discussion by schnaps91** [joonaspaakko/Batch-Mockup-Smart-Object-Replacement-photoshop-script#19] [Q&A]
First off, hello and thank you for this brilliant script, been trying to get one for months now to no avail, so thank you for saving up much of my time.
I would ask if it would be easy to modify it a tiny bit, to loop for a specified number of iterations while fetching random images from each input folder instead of going linearly until all files in the folder have been placed.

Thanks in advance

**Comment by joonaspaakko** [comment]
I don't quite understand the point of either of those. Wouldn't looping through the input files multiple times create duplicate output files? And the randomization would create the same output files just in a different order each time. 

The only possible use-case I can come up with _(but not really justify)_, is if you have multiple smart objects and you want every possible combination, but randomization wouldn't exactly guarantee that. It would have the same duplicate outputs issue, I think.

I didn't spend a long time thinking about it but I couldn't come up with a scenario where either make sense?

**Reply by schnaps91** [reply]
> I don't quite understand the point of either of those. Wouldn't looping through the input files multiple times create duplicate output files? And the randomization would create the same output files just in a different order each time.
> 
> The only possible use-case I can come up with_(but not really justify)_, is if you have multiple smart objects and you want every possible combination, but randomization wouldn't exactly guarantee that. It would have the same duplicate outputs issue, I think.
> 
> I didn't spend a long time thinking about it but I couldn't come up with a scenario where either make sense?

Well you partially guessed correctly, i do have a mockup with about 8 smart objects which have different aspect ratios. Considering the pool of inputs i have is in the triple digits, randomizing the output will save me quite a lot of time because i can run the script over and over and then loop over the outputs and choose from them. I hope that made sense

**Reply by joonaspaakko** [reply]
Would it be better to have an option that makes sure it outputs every single combination or does that undermine the point?

**Reply by schnaps91** [reply]
> Would it be better to have an option that makes sure it outputs every single combination or does that undermine the point?

It would certainly be the absolute best, however there would have to be a specified number of iterations since if the inputs are too many, outputting every possible combination would clog the disk.  Hence why i added the randomization because if it did it linearly it would output the same combinations each time before running out of disk space.

**Reply by joonaspaakko** [reply]
I think I understand now. I'll have to think about it.

**Reply by schnaps91** [reply]
Great! I think a way to go about it could be to make a string array with the filepaths to each input file and then have it pick one through a rand function but I am no coder so what i am saying might be completely wrong. Either way huge thanks for your work, it is amazing

**Reply by joonaspaakko** [reply]
[Here's what I came up with](https://gist.github.com/joonaspaakko/89d3b305f4a190a57d8f1800fbb48490/revisions). I linked straight to the revisions so you can see the code example at the top and also what changed from the original code. I made this change in a gist and not directly to this project. The "gist of it" is that if you imagine you have 1 mockup with 3 smart objects, it goes through every smart object input individually by first randomizing the input and then limiting each input array to the given length.

> When we last talked I had the flu, so it took me a while to actually take a look at this... It's not like I worked a week on this, I just basically went with the easiest possible thing that is incredibly flawed... But I thought I'd also mention the better way that I came up with, but didn't do because I deemed it too much work:
> 
> 1. Create a text file with all file paths. 
>      - Only create the file if it doesn't already exist. 
>      - Shuffle paths on creation if needed.
> 2. Pick and remove files from the text file on every run
> 
> This way it could systematically work through every single file a chunk at a time and you would never get duplicates.

**Reply by schnaps91** [reply]
> [Here's what I came up with](https://gist.github.com/joonaspaakko/89d3b305f4a190a57d8f1800fbb48490/revisions). I linked straight to the revisions so you can see the code example at the top and also what changed from the original code. I made this change in a gist and not directly to this project. The "gist of it" is that if you imagine you have 1 mockup with 3 smart objects, it goes through every smart object input individually by first randomizing the input and then limiting each input array to the given length.
> 
> > When we last talked I had the flu, so it took me a while to actually take a look at this... It's not like I worked a week on this, I just basically went with the easiest possible thing that is incredibly flawed... But I thought I'd also mention the better way that I came up with, but didn't do because I deemed it too much work:
> > 
> > 1. Create a text file with all file paths.
> >    
> >    * Only create the file if it doesn't already exist.
> >    * Shuffle paths on creation if needed.
> > 2. Pick and remove files from the text file on every run
> > 
> > This way it could systematically work through every single file a chunk at a time and you would never get duplicates.

Thanks a lot man, hope you are better from the flu. What does it do in terms of combinatoric output? It just randomizes the inputs i presume, but for how many iterations? The length specified?

**Reply by joonaspaakko** [reply]
It's determined by the two extra options for smart objects:

```js
inputRandom: true, // Randomize input files for this smart object at the beginning of the batch process
inputLength: 2, // Limit input files of this smart object to 2 
```

You can see an example use case [here](https://gist.github.com/joonaspaakko/89d3b305f4a190a57d8f1800fbb48490#file-batch-mockup-smart-object-replacement-jsx-L27-L28). And just to be clear, if you have like 4 smart objects that you want to limit and randomize, you put these two lines in all of them. Yea it randomizes the input files for each smart object separately, which in turn randomizes the output. 

> One potential pain point is the output filenames. Any duplicate output filenames will end up overwriting files that were output in the last batch. This is pretty much just how it is... I believe the only way to properly make sure it would never overwrite output files would be to read the output folder(s) to check how many output files there are currently and add incremental numbers to the new output files starting from there.
